src/neighbor.h

```diff
@@ -0,0 +1,85 @@
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+//************************************************************************
+//  ExaMiniMD v. 1.0
+//  Copyright (2018) National Technology & Engineering Solutions of Sandia,
+//  LLC (NTESS).
+//
+//  Under the terms of Contract DE-NA-0003525 with NTESS, the U.S. Government
+//  retains certain rights in this software.
+//
+//  ExaMiniMD is licensed under 3-clause BSD terms of use: Redistribution and
+//  use in source and binary forms, with or without modification, are
+//  permitted provided that the following conditions are met:
+//
+//    1. Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//
+//    2. Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//    3. Neither the name of the Corporation nor the names of the contributors
+//       may be used to endorse or promote products derived from this software
+//       without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY EXPRESS OR IMPLIED
+//  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+//  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+//  IN NO EVENT SHALL NTESS OR THE CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+//  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+//  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+//  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+//  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+//  IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+//  POSSIBILITY OF SUCH DAMAGE.
+//
+//************************************************************************
+
+#ifndef NEIGHBOR_H
+#define NEIGHBOR_H
+#include <system.h>
+#include <types.h>
+
+class Neighbor
+{
+  public:
+    T_X_FLOAT neigh_cut;
+    bool half_neigh;
+
+    virtual void init( T_X_FLOAT neigh_cut, bool half_neigh );
+    virtual void create( System *system );
+
+    template <class t_neigh_list>
+    t_neigh_list &get();
+
+    virtual const char *name();
+};
+
+template <class t_neigh_list>
+class NeighborVerlet : public Neighbor
+{
+  public:
+    t_neigh_list list;
+
+    void init( T_X_FLOAT neigh_cut, bool half_neigh );
+    void create( System *system );
+
+    t_neigh_list &get();
```

Comment:
When we add the tree class how will the type `t_neigh_list` be resolved? Will both the verlet and tree lists return different types? If so this lends itself to the force functions being templated on a neighbor list type.

Reply:
Ultimately I think we want to converge on a single Cabana neighbor list data type (independent of the construction method such as Verlet or Tree). Then you would return that type here and this function can also be pure virtual in the base class. In the short term I think this makes the most sense and we move the Cabana neighbor list traits interface into an interface in that simple class. In the future we will explore computing forces as we traverse the tree which would be a different paradigm.

Reply:
Agreed. Following from the layout PR, I plan to template both Neighbor and Force on the neighbor type, which would work fine with separate Cabana list types. But I think this current approach could work with Cabana list types that are covariant - when I tested adding the Tree class I saw that you're right, this doesn't work with multiple child classes

Reply:
This is taken care of, as you suggested, with a template over `Neighbor` types